src/v/storage/segment_utils.cc

```diff
@@ -407,16 +411,20 @@ ss::future<> do_self_compact_segment(
                 return do_copy_segment_data(s, cfg, pb, std::move(h));
             });
       })
-      .then([s](storage::index_state idx) {
-          return s->write_lock().then(
-            [s, idx = std::move(idx)](ss::rwlock::holder h) mutable {
-                using type = std::tuple<index_state, ss::rwlock::holder>;
-                if (s->is_closed()) {
-                    return ss::make_exception_future<type>(
-                      segment_closed_exception());
-                }
-                return ss::make_ready_future<type>(
-                  std::make_tuple(std::move(idx), std::move(h)));
+      .then([s, &readers_cache](storage::index_state idx) {
+          return readers_cache.evict_segment_readers(s).then(
+            [s,
+             idx = std::move(idx)](readers_cache::range_lock_holder) mutable {
```

Comment:
why is holder's lifetime not prolonged?

Reply:
we can relay on the fact that `write_lock` is synchronous up to the point where waiter is put to the `_waiters` list